hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -516,17 +505,52 @@ class MyStateMachine(RuleBasedStateMachine):
     drawn from this bundle will be consumed (as above) when requested.
     """
 
-    def __init__(self, name: str, *, consume: bool = False) -> None:
+    def __init__(self, name: str, *, consume: bool = False, transformations=()) -> None:
+        super().__init__(
+            (1, 2, 3), transformations=transformations
+        )  # Some random items that'll get replaced in do_draw
         self.name = name
-        self.__reference_strategy = BundleReferenceStrategy(name, consume=consume)
+        self.consume = consume
+
+        # Shrink towards the right rather than the left. This makes it easier
+        # to delete data generated earlier, as when the error is towards the
+        # end there can be a lot of hard to remove padding.
+        self.shrink_towards = sys.maxsize
+
+    def reference_to_val_func(self, dic, item):
+        assert isinstance(item, VarReference)
+        return dic.get(item.name)
 
     def do_draw(self, data):
         machine = data.draw(self_strategy)
-        reference = data.draw(self.__reference_strategy)
-        return machine.names_to_values[reference.name]
+
+        bundle = machine.bundle(self.name)
+        if not bundle:
+            data.mark_invalid(f"Cannot draw from empty bundle {self.name!r}")
+        self.elements = bundle
+        self.reference_to_value = partial(
+            self.reference_to_val_func, machine.names_to_values
+        )
+
+        ref = super().do_draw(data)
+        if self.consume:
+            bundle.remove(ref)  # pragma: no cover # coverage is flaky here
```

Comment:
I'd be inclined to restore our placeholder `self.elements` here.

I'm also wondering what happens if you have a bundle which holds a tree of some kind.  Is it possible to get an indexing error from recursive references if something is consumed at the wrong time?  I'll need to think about this; we might be fine or we might need something a touch more complicated for consumes bundles...

Reply:
I think there's also a latent bug here where `bundle.remove(ref)` can remove a different object to the one we selected, if they're equal, which could be a really subtle problem for someone.

The solution is probably to do `elements = range(len(bundle))[::-1]` and modify `reference_to_val_func` to also accept the `bundle` list and do a two-step lookup + record the index, and finally pop from that index.  Ouch.